splash/qtutils.py

```diff
@@ -254,171 +256,380 @@ def swap_byte_order_i32(buf):
     return arr.tostring()
 
 
+def render_qwebpage(web_page, logger=None, width=None, height=None,
+                    scale_method=None):
+    """
+    Render QWebPage into a WrappedImage.
+
+    :type web_page: PyQt4.QtWebKit.QWebPage
+    :type logger: splash.browser_tab._BrowserTabLogger
+    :type width: int
+    :type height: int
+    :type scale_method: str {'raster', 'vector'}
+    :rtype: WrappedImage
+
+    """
+    # Overall rendering pipeline looks as follows:
+    # 1. render_qwebpage
+    # 2. render_qwebpage_raster/-vector
+    # 3. render_qwebpage_impl
+    # 4. render_qwebpage_full/-tiled
+    if logger is None:
+        logger = _DummyLogger()
+    if scale_method is None:
+        scale_method = defaults.PNG_SCALE_METHOD
+
+    web_viewport = QRect(QPoint(0, 0), web_page.viewportSize())
+    img_viewport, img_size = _calculate_image_parameters(
+        web_viewport, width, height)
+    logger.log("png render: output size=%s, viewport=%s" %
+               (img_size, img_viewport), min_level=2)
+
+    if scale_method == 'vector':
+        return _render_qwebpage_vector(
+            web_page, logger,
+            in_viewport=web_viewport, out_viewport=img_viewport,
+            image_size=img_size)
+    elif scale_method == 'raster':
+        return _render_qwebpage_raster(
+            web_page, logger,
+            in_viewport=web_viewport, out_viewport=img_viewport,
+            image_size=img_size)
+    else:
+        raise ValueError(
+            "Invalid scale method (must be 'vector' or 'raster'): %s" %
+            str(scale_method))
+
+
+def _render_qwebpage_vector(web_page, logger,
+                            in_viewport, out_viewport, image_size):
+    """
+    Render a webpage using vector rescale method.
+
+    :param in_viewport: region of the webpage to render from
+    :param out_viewport: region of the image to render to
+    :param image_size: size of the resulting image
+
+    :type in_viewport: QRect
+    :type out_viewport: QRect
+    :type image_size: QSize
+
+    """
+    web_rect = QRect(in_viewport)
+    render_rect = QRect(out_viewport)
+    canvas_size = QSize(image_size)
+
+    logger.log("png render: rendering %s of the web page" % web_rect,
+               min_level=2)
+    logger.log("png render: rendering into %s of the canvas" % render_rect,
+               min_level=2)
+    logger.log("png render: canvas size=%s" % canvas_size, min_level=2)
+
+    to_paint = render_rect.intersected(QRect(QPoint(0, 0), canvas_size))
+    # QPainter has several issues when its rendering area is more than 2**15 in
+    # any dimension.
+    need_tiling = max(to_paint.width(), to_paint.height()) >= (1 << 15)
```

Comment:
`2**15` reads better :)

Also, the comment says "more than", while the code uses ">=". I don't care about it though, ">=" is fine to be on a safe side.